Windows support

User mpilgrem on reddit reports:

Thank you drwebb and mboes. I tried `stack --extra-lib-dirs "C:\Program Files\R\R-3.2.2\bin\x64" --extra-include-dirs "C:\Program Files\R\R-3.2.2\include" build --copy-bins H`

but it failed with a different error while building package inline-r-0.7.0.0:
```
Configuring inline-r-0.7.0.0...
Building inline-r-0.7.0.0...
Preprocessing library inline-r-0.7.0.0...
c2hs.exe: C header contains errors:

C:\\PROGRA~1\\R\\R-3.2.2\\include/R_ext/eventloop.h:73: (column 65) [ERROR]  >>> Syntax error !
The symbol `fd_set' does not fit here.
```
Line 73 of eventloop.h reads:
```
extern InputHandler *getSelectedHandler(InputHandler *handlers, fd_set *mask);
```
That's a dead end for me. I'll wait for Windows support.

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,0 +1,42 @@
+version: "{build}"
+cache: C:\Users\appveyor\AppData\Roaming\stack
+
+build:
+  verbosity: normal
+
+deploy: off
+
+install:
+  ps: |
+    bash -c 'curl -Lso stack.zip -L --insecure http://www.stackage.org/stack/windows-i386'
+    7z x stack.zip stack.exe
+    mkdir C:\stack
+    mv stack.exe C:\stack
+
+    Write-Host "Downloading R.vhd"
+    bash -c 'curl -Ls https://rportable.blob.core.windows.net/r-portable/master/R.vhd.gz | gunzip -c > ../R.vhd'
+
+    Write-Host "Getting full path for R.vhd"
+    $ImageFullPath = Get-ChildItem "..\R.vhd" | % { $_.FullName }
+    $ImageSize = (Get-Item $ImageFullPath).length
+    echo "$ImageFullPath [$ImageSize bytes]"
+
+    Write-Host "Mounting R.vhd"
+    $RDrive = [string](Mount-DiskImage -ImagePath $ImageFullPath -Passthru | Get-DiskImage | Get-Disk | Get-Partition | Get-Volume).DriveLetter + ":"
+    # Assert that R was mounted properly
+    if ( -not (Test-Path "${RDrive}\R\bin" -PathType Container) ) {
+      Throw "Failed to mount R. Could not find directory: ${RDrive}\R\bin"
+    }
+    echo "R is now available on drive $RDrive"
+
+    Write-Host "Setting PATH"
+    $env:PATH = $RDrive + '\Rtools\bin;' + $RDrive + '\Rtools\MinGW\bin;' + $RDrive + '\Rtools\gcc-4.6.3\bin;' + $RDrive + '\R\bin\i386;' + 'C:\stack;' + $env:PATH
+    $env:PATH.Split(";")
+
+    Write-Host "Setting R_LIBS_USER"
+    $env:R_LIBS_USER = 'c:\RLibrary'
+    mkdir $env:R_LIBS_USER
+
+build_script:
+  - stack --no-terminal setup > nul
+  - stack --no-terminal build --extra-lib-dirs=D:\R\bin\i386 --extra-include-dirs=D:\R\include inline-r H

--- a/inline-r/inline-r.cabal
+++ b/inline-r/inline-r.cabal
@@ -48,7 +48,6 @@ library
                        Foreign.R
                        Foreign.R.Constraints
                        Foreign.R.Embedded
-                       Foreign.R.EventLoop
                        Foreign.R.Error
                        Foreign.R.Parse
                        Foreign.R.Type
@@ -56,7 +55,6 @@ library
                        H.Prelude.Interactive
                        Language.R
                        Language.R.Debug
-                       Language.R.Event
                        Language.R.GC
                        Language.R.Globals
                        Language.R.HExp
@@ -66,6 +64,9 @@ library
                        Language.R.Internal
                        Language.R.Internal.FunWrappers
                        Language.R.Internal.FunWrappers.TH
+  if !os(windows)
+    exposed-modules:   Foreign.R.EventLoop
+                       Language.R.Event
   other-modules:       Control.Monad.R.Class
                        Control.Monad.R.Internal
                        Data.Vector.SEXP.Mutable.Internal

--- a/inline-r/inline-r.cabal
+++ b/inline-r/inline-r.cabal
@@ -104,11 +104,6 @@ library
   else
     build-depends:     unix >= 2.6
     pkgconfig-depends: libR >= 3.0
-    cpp-options:       -DH_ARCH_UNIX
-    cc-options:        -DH_ARCH_UNIX
-    if os(darwin)
-      cpp-options:     -DH_ARCH_UNIX_DARWIN
-      cc-options:      -DH_ARCH_UNIX_DARWIN
   -- XXX -fcontext-stack=32 required on GHC >= 7.8 to allow foreign
   -- export function -wrappers of high arities.
   ghc-options:         -Wall -fcontext-stack=32

--- a/inline-r/inline-r.cabal
+++ b/inline-r/inline-r.cabal
@@ -101,8 +101,6 @@ library
                      , hsc2hs
   if os(windows)
     extra-libraries:   R
-    cpp-options:       -DH_ARCH_WINDOWS
-    cc-options:        -DH_ARCH_WINDOWS
   else
     build-depends:     unix >= 2.6
     pkgconfig-depends: libR >= 3.0

--- a/inline-r/src/Foreign/R.chs
+++ b/inline-r/src/Foreign/R.chs
@@ -174,7 +174,6 @@ import Prelude hiding (asTypeOf, length)
 
 #define USE_RINTERNALS
 #include <R.h>
-#include <Rinterface.h>
 #include <Rinternals.h>
 #include <R_ext/Memory.h>
 #include "missing_r.h"

--- a/inline-r/src/Foreign/R.chs
+++ b/inline-r/src/Foreign/R.chs
@@ -166,9 +166,6 @@ import Data.Singletons (fromSing)
 import Data.Typeable (Typeable)
 #endif
 import Foreign (Ptr, castPtr, plusPtr, Storable(..))
-#ifdef H_ARCH_WINDOWS
-import Foreign (nullPtr)
-#endif
 import Foreign.C
 import Prelude hiding (asTypeOf, length)
 

--- a/inline-r/src/H/Prelude.hs
+++ b/inline-r/src/H/Prelude.hs
@@ -11,7 +11,10 @@ module H.Prelude
   , module Foreign.R.Error
   -- * Language.R functions
   , module Language.R
+-- Not supported on Windows.
+#ifndef mingw32_HOST_OS
   , module Language.R.Event
+#endif
   , module Language.R.HExp
   , module Language.R.Literal
   , module Language.R.QQ
@@ -24,7 +27,9 @@ import           Language.R.HExp
 
 -- Reexported modules.
 import           Language.R hiding (SEXPTYPE(..))
+#ifndef mingw32_HOST_OS
 import           Language.R.Event (refresh)
+#endif
 import           Language.R.Globals
 import           Language.R.Instance
 import           Language.R.Literal

--- a/inline-r/src/Language/R/Globals.hs
+++ b/inline-r/src/Language/R/Globals.hs
@@ -18,8 +18,10 @@ module Language.R.Globals
   , unboundValue
   -- * R Internal constants
   , isRInteractive
-  , inputHandlers
   , signalHandlersPtr
+#ifndef mingw32_HOST_OS
+  , inputHandlers
+#endif
   -- * R global constants
   -- $ghci-bug
   , pokeRVariables
@@ -38,7 +40,9 @@ import Foreign
 import Foreign.C.Types (CInt)
 import Foreign.R (SEXP)
 import qualified Foreign.R as R
+#ifndef mingw32_HOST_OS
 import qualified Foreign.R.EventLoop as R
+#endif
 import System.IO.Unsafe (unsafePerformIO)
 
 -- $ghci-bug
@@ -59,8 +63,10 @@ type RVariables =
     , Ptr (SEXP G 'R.Symbol)
     , Ptr (SEXP G 'R.Symbol)
     , Ptr CInt
-    , Ptr (Ptr R.InputHandler)
     , Ptr CInt
+#ifndef mingw32_HOST_OS
+    , Ptr (Ptr R.InputHandler)
+#endif
     )
 
 -- | Stores R variables in a static location. This makes the variables'
@@ -77,8 +83,10 @@ pokeRVariables = poke rVariables <=< newStablePtr
  , unboundValuePtr
  , missingArgPtr
  , isRInteractive
- , inputHandlersPtr
  , signalHandlersPtr
+#ifndef mingw32_HOST_OS
+ , inputHandlersPtr
+#endif
  ) = unsafePerformIO $ peek rVariables >>= deRefStablePtr
 
 -- | Special value to which all symbols unbound in the current environment
@@ -106,5 +114,7 @@ emptyEnv = unsafePerformIO $ peek emptyEnvPtr
 globalEnv :: SEXP G 'R.Env
 globalEnv = unsafePerformIO $ peek globalEnvPtr
 
+#ifndef mingw32_HOST_OS
 inputHandlers :: Ptr R.InputHandler
 inputHandlers = unsafePerformIO $ peek inputHandlersPtr
+#endif

--- a/inline-r/src/Language/R/Instance.hs
+++ b/inline-r/src/Language/R/Instance.hs
@@ -44,7 +44,9 @@ import           Data.Monoid
 import           Data.Default.Class (Default(..))
 import qualified Foreign.R as R
 import qualified Foreign.R.Embedded as R
+#ifndef mingw32_HOST_OS
 import qualified Foreign.R.EventLoop as R
+#endif
 import           Foreign.C.String
 import           Language.R.Globals
 
@@ -240,8 +242,10 @@ initialize Config{..} = do
           , R.unboundValue
           , R.missingArg
           , R.isRInteractive
-          , R.inputHandlers
           , R.signalHandlers
+#ifndef mingw32_HOST_OS
+          , R.inputHandlers
+#endif
           )
         populateEnv
         args <- (:) <$> maybe getProgName return (getLast configProgName)

--- a/inline-r/src/Language/R/Instance.hs
+++ b/inline-r/src/Language/R/Instance.hs
@@ -73,7 +73,7 @@ import System.Environment ( getProgName, lookupEnv )
 import System.IO.Unsafe   ( unsafePerformIO )
 import System.Process     ( readProcess )
 import System.SetEnv
-#ifdef H_ARCH_UNIX
+#ifndef mingw32_HOST_OS
 import Control.Exception ( onException )
 import System.IO ( hPutStrLn, stderr )
 import System.Posix.Resource
@@ -205,8 +205,8 @@ initLock = unsafePerformIO $ newMVar ()
 -- to achieve this.
 initialize :: Config -> IO ()
 initialize Config{..} = do
-#ifdef H_ARCH_UNIX
-#ifdef H_ARCH_UNIX_DARWIN
+#ifndef mingw32_HOST_OS
+#ifdef darwin_HOST_OS
     -- NOTE: OS X does not allow removing the stack size limit completely,
     -- instead forcing a hard limit of just under 64MB.
     let stackLimit = ResourceLimit 67104768
@@ -218,7 +218,7 @@ initialize Config{..} = do
                        "Language.R.Interpreter: "
                        ++ "Cannot increase stack size limit."
                        ++ "Try increasing your stack size limit manually:"
-#ifdef H_ARCH_UNIX_DARWIN
+#ifdef darwin_HOST_OS
                        ++ "$ launchctl limit stack 67104768"
                        ++ "$ ulimit -s 65532"
 #else

